File: step-api-net/StepApiKeyword/Keyword.cs
using Newtonsoft.Json.Linq;
using Step.Functions.IO;
using System;
using System.Collections.Generic;
using System.Reflection;

namespace Step.Handlers.NetHandler
{
    [AttributeUsage(AttributeTargets.Method)]
    public class Keyword : Attribute
    {
        /// <summary>
        /// Reserved keyword for the routing element; execution will occur on the controller rather than on agents
        /// </summary>
        public const string ROUTING_EXECUTE_ON_CONTROLLER = "controller";
        
        public string name;
        public string description;
        public string schema;
        public string[] properties;
        public string[] optionalProperties;
        // keyword timeout in milliseconds:
        public int timeout = 180000;

        /// <summary>
        /// Routing criteria. Can be:
        /// - empty: use default routing
        /// - <see cref="Keyword.ROUTING_EXECUTE_ON_CONTROLLER"/>: execute on controller rather than on agents
        /// - array of key-value pairs for agent token selection criteria
        /// </summary>
        public string[] routing;
    }

    public class AbstractKeyword
    {
        public JObject input;

        public Dictionary<string, string> properties;

        public OutputBuilder output;

        public TokenSession session;

        public TokenSession tokenSession;

        /// <summary>
        /// Hook called when an exception is thrown by a keyword or by the BeforeKeyword hooks
        /// </summary>
        /// <param name="e">the exception thrown</param>
        /// <returns>return true if the exception passed as argument has to be rethrown</returns>
        public virtual bool OnError(Exception e)
        {
            return true;
        }

        /// <summary>
        /// Hook called before each keyword call.
        /// If an exception is thrown by this method, the keyword won't be executed 
        /// (but AfterKeyword and OnError will)
        /// </summary>
        /// <param name="KeywordName">the name of the keyword.  Will be the function name if annotation.name is empty </param>
        /// <param name="Annotation">the annotation of the called keyword</param>
        public virtual void BeforeKeyword(String KeywordName, Keyword Annotation) { }

        /// <summary>
        /// Hook called after each keyword call.
        /// This method is always called. If an exception is thrown by the keyword or the BeforeKeyword hook,
        /// this method is called after the OnError hook.
        /// </summary>
        /// <param name="KeywordName">the name of the keyword.  Will be the function name if annotation.name is empty </param>
        /// <param name="Annotation">the annotation of the called keyword</param>
        public virtual void AfterKeyword(String KeywordName, Keyword Annotation) { }
    }
}

File: step-api/step-api-keyword/src/main/java/step/handlers/javahandler/Keyword.java
/*******************************************************************************
 * Copyright (C) 2020, exense GmbH
 *  
 * This file is part of STEP
 *  
 * STEP is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *  
 * STEP is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *  
 * You should have received a copy of the GNU Affero General Public License
 * along with STEP.  If not, see <http://www.gnu.org/licenses/>.
 ******************************************************************************/
package step.handlers.javahandler;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
public @interface Keyword {

	/** Reserved keyword for the routing element; execution will occur on the controller rather than on agents */
    String EXECUTE_ON_CONTROLLER = "controller";
	/**
	 * @return the name of this keyword. If not specified the method name is 
	 * used as keyword name
	 */
	String name() default "";

	/**
	 * @return the description of this keyword.
	 */
	String description() default "";
	
	/**
	 * @return the list of properties required by this keyword
	 */
	String[] properties() default {};
	
	/**
	 * @return the list of optional properties which might be used by this keyword
	 */
	String[] optionalProperties() default {};
	
	/**
	 * @return the JSON schema of the input object. 
	 * @see <a href="https://json-schema.org/"> json-schema.org </a>
	 */
	String schema() default "";

	/**
	 * @return the call timeout for this keyword in milliseconds
	 */
	int timeout() default 180000;

	/**
	 * @return the reference to file with implementing plan (for composite keyword only)
	 */
	String planReference() default "";

	/**
	 *
	 * @return the routing criteria which can be either <ul>
	 *     <li>empty: use default routing on Step</li>
	 *     <li>{@link #EXECUTE_ON_CONTROLLER}: execute the keyword directly on the controller rather than on agents</li>
	 *     <li>"agent_attribute_key_1","agent_attribute_value_1",...: the agent token selection criteria provided as a list of key-value pairs</li>
	 * </ul>
	 */
	String[] routing() default {};
}
